File: src/ast/CvType.java
package ast;

import java.io.StringWriter;

public class CvType extends ParentNode {

    public class Cv {
        public static final int None = 1;
        public static final int Const = 2;
        public static final int Volatile = 3;
        public static final int Restricted = 4;
    }

    public int Qualifier;

    public CvType(int qualifier, BaseNode child) {
        super(NodeType.CvQualifierType, child);
        Qualifier = qualifier;
    }

    public void PrintQualifier(StringWriter writer) {
        if ((Qualifier & Cv.Const) != 0) {
            writer.write(" const");
        }

        if ((Qualifier & Cv.Volatile) != 0) {
            writer.write(" volatile");
        }

        if ((Qualifier & Cv.Restricted) != 0) {
            writer.write(" restrict");
        }
    }

    @Override
    public void PrintLeft(StringWriter writer) {
        if (child != null) {
            child.PrintLeft(writer);
        }

        PrintQualifier(writer);
    }

    @Override
    public boolean HasRightPart() {
        return child != null && child.HasRightPart();
    }

    @Override
    public void PrintRight(StringWriter writer) {
        if (child != null) {
            child.PrintRight(writer);
        }
    }
}

File: src/de/fearlesstobi/demangler/ast/CvType.kt
package de.fearlesstobi.demangler.ast

import java.io.StringWriter

class CvType(private val qualifier: Int, child: BaseNode?) : ParentNode(NodeType.CvQualifierType, child) {
    object Cv {
        const val None = 0
        const val Const = 1
        const val Volatile = 2
        const val Restricted = 4
    }

    private fun PrintQualifier(writer: StringWriter) {
        if (qualifier and Cv.Const != 0) {
            writer.write(" const")
        }
        if (qualifier and Cv.Volatile != 0) {
            writer.write(" volatile")
        }
        if (qualifier and Cv.Restricted != 0) {
            writer.write(" restrict")
        }
    }

    public override fun printLeft(writer: StringWriter) {
        child?.printLeft(writer)
        PrintQualifier(writer)
    }

    public override fun hasRightPart(): Boolean {
        return child != null && child.hasRightPart()
    }

    public override fun printRight(writer: StringWriter) {
        child?.printRight(writer)
    }

}